Firmware/RTK_Everywhere/Form.ino

```diff
@@ -11,49 +11,239 @@ Form.ino
+// https://github.com/espressif/arduino-esp32/blob/master/libraries/ESP32/examples/Camera/CameraWebServer/CameraWebServer.ino
+// https://esp32.com/viewtopic.php?t=24445
+
+// These are useful:
+// https://github.com/mo-thunderz/Esp32WifiPart2/blob/main/Arduino/ESP32WebserverWebsocket/ESP32WebserverWebsocket.ino
+// https://www.youtube.com/watch?v=15X0WvGaVg8
+
+void sendStringToWebsocket(const char *stringToSend)
+{
+    if (!websocketConnected)
+    {
+        systemPrintf("sendStringToWebsocket: not connected - could not send: %s\r\n", stringToSend);
+        return;
+    }
+
+    // To send content to the webserver, we would call: webserver->sendContent(stringToSend);
+    //
+    // But here we want to send content to the websocket (wsserver)...
+
+    httpd_ws_frame_t ws_pkt;
+    memset(&ws_pkt, 0, sizeof(httpd_ws_frame_t));
+    ws_pkt.payload = (uint8_t*)stringToSend;
+    ws_pkt.len = strlen(stringToSend);
+    ws_pkt.type = HTTPD_WS_TYPE_TEXT;
+
+    // If we use httpd_ws_send_frame, it requires a req.
+    //esp_err_t ret = httpd_ws_send_frame(last_ws_req, &ws_pkt);
+    //if (ret != ESP_OK) {
+    //    ESP_LOGE(TAG, "httpd_ws_send_frame failed with %d", ret);
+    //}
+
+    // If we use httpd_ws_send_frame_async, it requires a fd.
+    esp_err_t ret = httpd_ws_send_frame_async(*wsserver, last_ws_fd, &ws_pkt);
+    if (ret != ESP_OK) {
+        systemPrintf("httpd_ws_send_frame failed with %d\r\n", ret);
+    }
+    else
+    {
+        if (settings.debugWebConfig == true)
+            systemPrintf("sendStringToWebsocket: %s\r\n", stringToSend);
+    }
+
+}
+
+static esp_err_t ws_handler(httpd_req_t *req)
+{
+    // Log the req, so we can reuse it for httpd_ws_send_frame
+    // TODO: do we need to be cleverer about this?
+    //last_ws_req = req;
+
+    if (req->method == HTTP_GET) {
+        // Log the fd, so we can reuse it for httpd_ws_send_frame_async
+        // TODO: do we need to be cleverer about this?
+        last_ws_fd = httpd_req_to_sockfd(req);
+
+        if (settings.debugWebConfig == true)
+            systemPrintf("Handshake done, the new ws connection was opened with fd %d\r\n", last_ws_fd);
+
+        websocketConnected = true;
+        lastDynamicDataUpdate = millis();
+        sendStringToWebsocket(settingsCSV);
+
+        return ESP_OK;
+    }
+
+    httpd_ws_frame_t ws_pkt;
+    uint8_t *buf = NULL;
+    memset(&ws_pkt, 0, sizeof(httpd_ws_frame_t));
+    ws_pkt.type = HTTPD_WS_TYPE_TEXT;
+    /* Set max_len = 0 to get the frame len */
+    esp_err_t ret = httpd_ws_recv_frame(req, &ws_pkt, 0);
+    if (ret != ESP_OK) {
+        systemPrintf("httpd_ws_recv_frame failed to get frame len with %d\r\n", ret);
+        return ret;
+    }
+    if (settings.debugWebConfig == true)
+        systemPrintf("frame len is %d\r\n", ws_pkt.len);
+    if (ws_pkt.len) {
+        /* ws_pkt.len + 1 is for NULL termination as we are expecting a string */
+        buf = (uint8_t *)calloc(1, ws_pkt.len + 1);
```

Comment:
Should you be calloc'ing to PSRAM? It's probably not much but could be done as a rule of thumb. Search codebase for malloc to see how we handle psram online/offline.

Reply:
Done - thanks! : 8cdcc2be6cee8a8db0d36c90d078c4b316112ff5